website/docs/language/tests/index.mdx

```diff
@@ -0,0 +1,685 @@
+    error_message = "invalid value for secondary S3 bucket"
+  }
+}
+```
+
+It is also possible to define specific providers you want to use in specific `run` blocks:
+
+```hcl
+# main.tf
+
+terraform {
+  required_providers {
+    aws = {
+      source                = "hashicorp/aws"
+      configuration_aliases = [aws.secondary]
+    }
+  }
+}
+
+data "aws_region" "primary" {}
+
+data "aws_region" "secondary" {
+  provider = aws.secondary
+}
+
+variable "bucket_prefix" {
+  default = "test"
+  type    = string
+}
+
+resource "aws_s3_bucket" "primary_bucket" {
+  bucket = "${var.bucket_prefix}-${data.aws_region.primary.name}-primary"
+}
+
+resource "aws_s3_bucket" "secondary_bucket" {
+  provider = aws.secondary
+  bucket   = "${var.bucket_prefix}-${data.aws_region.secondary.name}-secondary"
+}
+```
+
+```hcl
+provider "aws" {
+  region = "us-east-1"
+}
+
+provider "aws" {
+  alias  = "secondary"
+  region = "eu-central-1"
+}
+
+provider "aws" {
+  alias  = "tertiary"
+  region = "eu-west-2"
+}
+
+run "providers" {
+
+  command = plan
+
+  assert {
+    condition     = aws_s3_bucket.primary_bucket.bucket == "test-us-east-1-primary"
+    error_message = "invalid value for primary S3 bucket"
+  }
+
+  assert {
+    condition     = aws_s3_bucket.secondary_bucket.bucket == "test-eu-central-1-secondary"
+    error_message = "invalid value for secondary S3 bucket"
+  }
+}
+
+run "providers" {
+
+  command = plan
+
+  providers = {
+    aws           = aws
+    aws.secondary = aws.tertiary
+  }
+
+  assert {
+    condition     = aws_s3_bucket.primary_bucket.bucket == "test-us-east-1-primary"
+    error_message = "invalid value for primary S3 bucket"
+  }
+
+  assert {
+    condition     = aws_s3_bucket.secondary_bucket.bucket == "test-eu-west-2-secondary"
+    error_message = "invalid value for secondary S3 bucket"
+  }
+}
+```
+
+> **Note:** When running tests with `command = apply` switching providers between `run` blocks can result in failed operations and tests as resources created by one provider definition will be unusable when modified by a second.
+
+## Modules
+
+You can also modify the module that a given `run` block will execute.
+
+By default, Terraform will execute the given command against the configuration under test for each `run` block. The configuration under test is the configuration within the directory the `terraform test` command was executed from (or the directory pointed to by `-chdir` argument). Each `run` block also allows the user to change the targeted configuration using the `module` block.
+
+Compared with the traditional [`module` block](/terraform/language/modules/syntax), the `module` block within test files only supports the [`source`](/terraform/language/modules/syntax#source) attribute and the [`version`](/terraform/language/modules/syntax#version) attribute. The remaining attributes that would normally be supplied via the traditional `module` block are provided elsewhere within the `run` block.
```

Comment:
It would be the user.

How about this?

```

Unlike the traditional [`module` block](/terraform/language/modules/syntax), the `module` block within test files _only_ supports the [`source`](/terraform/language/modules/syntax#source) attribute and the [`version`](/terraform/language/modules/syntax#version) attribute. The remaining attributes that are typically supplied via the traditional `module` block should be supplied by the alternate attributes and blocks within the `run` block.
```